bytecomp/bytepackager.ml

```diff
@@ -65,32 +65,51 @@ let rename_relocation packagename objfile mapping base (rel, ofs) =
   let rel' =
     match rel with
       Reloc_getglobal id ->
+        let id = Ident.create_persistent id in
         begin try
           let id', defined = Ident.Map.find id mapping in
           if defined
-          then Reloc_getglobal id'
+          then Reloc_getglobal (Ident.name id')
           else raise(Error(Forward_reference(objfile, id)))
         with Not_found ->
           (* PR#5276: unique-ize dotted global names, which appear
              if one of the units being consolidated is itself a packed
              module. *)
           let name = Ident.name id in
           if String.contains name '.' then
-            Reloc_getglobal (Ident.create_persistent (packagename ^ "." ^ name))
+            Reloc_getglobal (packagename ^ "." ^ name)
+          else
+            rel
+        end
+    | Reloc_getpredef id ->
+        let id = Ident.create_persistent id in
+        begin try
+          let id', defined = Ident.Map.find id mapping in
+          if defined
+          then Reloc_getpredef (Ident.name id')
+          else raise(Error(Forward_reference(objfile, id)))
+        with Not_found ->
+          (* PR#5276: unique-ize dotted global names, which appear
+             if one of the units being consolidated is itself a packed
+             module. *)
+          let name = Ident.name id in
+          if String.contains name '.' then
+            Reloc_getpredef (packagename ^ "." ^ name)
           else
             rel
         end
```

Comment:
There is a fair amount of code duplication between the `Reloc_getglobal` and the `Reloc_getpredef` cases above. Could we have the following instead?

```
| Reloc_getglobal of { predef: bool; id : global }
```